OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnCustomLayer.cs

```diff
@@ -0,0 +1,89 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class GrantConditionOnCustomLayerInfo : ITraitInfo
+	{
+		[GrantedConditionReference]
+		[Desc("The condition to grant to self while inside a tunnel.")]
+		public readonly string TunnelCondition = null;
+
+		[GrantedConditionReference]
+		[Desc("The condition to grant to self while underground.")]
+		public readonly string SubterraneanCondition = null;
+
+		[GrantedConditionReference]
+		[Desc("The condition to grant to self while flying.")]
+		public readonly string JumpjetCondition = null;
+
+		public object Create(ActorInitializer init) { return new GrantConditionOnCustomLayer(init, this); }
+	}
+
+	public class GrantConditionOnCustomLayer : INotifyCreated, INotifyChangedCustomLayer
```

Comment:
and then split this into separate traits for Tunnel/Subterranean/Jumpjet.  You could define a single abstract class that is given the `CustomLayerMovementType` to check in its ctor from the subclass.  The `CustomLayerChanged` method in each can then check this value and grant/revoke the condition when needed.